src/chain/stateTransition/epoch/util.ts

```diff
@@ -72,51 +71,58 @@ export function getAttestingBalance(state: BeaconState, attestations: PendingAtt
-      data.targetEpoch,
-      state.currentCrosslinks[data.shard].epoch + MAX_CROSSLINK_EPOCHS
-    ),
-    previousCrosslinkRoot: data.previousCrosslinkRoot,
-    crosslinkDataRoot: data.crosslinkDataRoot,
-  };
-}
+//SPEC 0.7
+// def get_winning_crosslink_and_attesting_indices(state: BeaconState,
+//   epoch: Epoch,
+//   shard: Shard) -> Tuple[Crosslink, List[ValidatorIndex]]:
+// attestations = [a for a in get_matching_source_attestations(state, epoch) if a.data.crosslink.shard == shard]
+// crosslinks = list(filter(
+//   lambda c: hash_tree_root(state.current_crosslinks[shard]) in (c.parent_root, hash_tree_root(c)),
+//   [a.data.crosslink for a in attestations]
+// ))
+// # Winning crosslink has the crosslink data root with the most balance voting for it (ties broken lexicographically)
+// winning_crosslink = max(crosslinks, key=lambda c: (
+//   get_attesting_balance(state, [a for a in attestations if a.data.crosslink == c]), c.data_root
+// ), default=Crosslink())
+// winning_attestations = [a for a in attestations if a.data.crosslink == winning_crosslink]
+// return winning_crosslink, get_unslashed_attesting_indices(state, winning_attestations)
+
 
 export function getWinningCrosslinkAndAttestingIndices(
   state: BeaconState,
   epoch: Epoch,
   shard: Shard
 ): [Crosslink, ValidatorIndex[]] {
-  const shardAttestations = getMatchingSourceAttestations(state, epoch)
-    .filter((a) => a.data.shard === shard);
-  const shardCrosslinks = shardAttestations
-    .map((a) => getCrosslinkFromAttestationData(state, a.data));
-  const currentCrosslinkRoot = hashTreeRoot(state.currentCrosslinks[shard], Crosslink); 
-  const candidateCrosslinks = shardCrosslinks.filter((c) => (
-    currentCrosslinkRoot.equals(c.previousCrosslinkRoot) ||
+
+  const attestations = getMatchingSourceAttestations(state, epoch)
+    .filter((a) => a.data.crosslink.shard === shard);
+
+  let output: Crosslink[];
+  attestations.forEach((a) => output.push(a.data.crosslink));
+
+  const currentCrosslinkRoot = hashTreeRoot(state.currentCrosslinks[shard], Crosslink);
+  const crosslinks = output.filter((c) => (
+    currentCrosslinkRoot.equals(c.parentRoot) ||
     currentCrosslinkRoot.equals(hashTreeRoot(c, Crosslink))
   ));
 
-  if (candidateCrosslinks.length === 0) {
-    return [{
-      epoch: GENESIS_EPOCH,
-      previousCrosslinkRoot: ZERO_HASH,
-      crosslinkDataRoot: ZERO_HASH,
-    }, []];
+  const defultCrossLink: Crosslink= {
+    shard: GENESIS_START_SHARD,
+    startEpoch: GENESIS_EPOCH,
+    endEpoch: GENESIS_EPOCH,
+    parentRoot: ZERO_HASH,
+    dataRoot: ZERO_HASH,
+  };
+
+  if (crosslinks.length === 0) {
+    return [defultCrossLink, []];
   }
-  
   const getAttestationsFor = (crosslink: Crosslink) => 
-    shardAttestations.filter((a) =>
-      serialize(getCrosslinkFromAttestationData(state, a.data), Crosslink)
-        .equals(serialize(crosslink, Crosslink)));
+    attestations.filter((a) => a.data.crosslink === crosslink);
```

Comment:
You cannot compare objects like that. For now serialize it using ssz and comapre buffers. When released, new ssz will have method equals for those objects